src/test/java/co/rsk/federate/FederationWatcherTest.java

```diff
@@ -22,354 +22,342 @@
             return null;
-        }).when(ethereumMock).addListener(any());
+        }).when(ethereum).addListener(any());
 
-        watcher.setup(federationProvider);
-        verify(ethereumMock).addListener(any());
+        // Act
+        federationWatcher.setup(federationProvider);
+
+        // Assert
+        verify(ethereum).addListener(any());
+        assertSame(TestUtils.getInternalState(federationWatcher, "federationProvider"), federationProvider);
     }
 
     @Test
-    void triggersActiveFederationChange_none_to_active() throws Exception {
-        EthereumListenerAdapter rskListener = setupAndGetRskListener(Optional.empty(), Optional.empty());
-        class EventsLogger
-        {
-            public int activeCalls = 0;
-            public int retiringCalls = 0;
-        }
-
-        when(federationProvider.getActiveFederationAddress()).thenReturn(federation1.getAddress());
-        when(federationProvider.getActiveFederation()).thenReturn(federation1);
+    void whenNoActiveFederation_shouldTriggerActiveFederationChange() throws Exception {
+        // Arrange
+        var rskListener = setupAndGetRskListener(Optional.empty(), Optional.empty());
+        var activeCalls = new AtomicInteger(0);
+        var retiringCalls = new AtomicInteger(0);
+
+        when(federationProvider.getActiveFederationAddress()).thenReturn(FIRST_FEDERATION.getAddress());
+        when(federationProvider.getActiveFederation()).thenReturn(FIRST_FEDERATION);
         when(federationProvider.getRetiringFederationAddress()).thenReturn(Optional.empty());
 
-        EventsLogger logger = new EventsLogger();
-
-        for (int i = 0; i < 2; i++) {
-            watcher.addListener(new FederationWatcher.Listener() {
-                @Override
-                public void onActiveFederationChange(Optional<Federation> oldFederation, Federation newFederation) {
-                    assertEquals(Optional.empty(), oldFederation);
-                    assertEquals(federation1, newFederation);
-                    logger.activeCalls++;
-                }
-
-                @Override
-                public void onRetiringFederationChange(Optional<Federation> oldFederation, Optional<Federation> newFederation) {
-                    logger.retiringCalls++;
-                }
-            });
-        }
-
+        federationWatcher.addListener(new FederationWatcher.Listener() {
+            @Override
+            public void onActiveFederationChange(Optional<Federation> oldFederation, Federation newFederation) {
+                assertEquals(Optional.empty(), oldFederation);
+                assertEquals(FIRST_FEDERATION, newFederation);
+                activeCalls.incrementAndGet();
+            }
+
+            @Override
+            public void onRetiringFederationChange(Optional<Federation> oldFederation, Optional<Federation> newFederation) {
+                retiringCalls.incrementAndGet();
+            }
+        });
```

Comment:
wdyt about extracting this to a method? I see it duplicated a couple of times

Reply:
For each test the anoymous class implementation is a bit different if you notice